plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpLocalRunTab.java

```diff
@@ -0,0 +1,290 @@
+          boolean savedEmailAvailable = accountSelector.isEmailAvailable(accountEmailModel);
+          if (!emptySelection || savedEmailAvailable) {
+            accountEmailModel = accountSelector.getSelectedEmail();
+            gcpProjectIdModel = ""; //$NON-NLS-1$
+            updateLaunchConfigurationDialog();
+          }
+        }
+      }
+    });
+
+    // Project row
+    Label projectLabel = new Label(composite, SWT.LEAD);
+    projectLabel.setText(Messages.getString("label.project")); //$NON-NLS-1$
+
+    Composite projectSelectorComposite = new Composite(composite, SWT.NONE);
+    final Text filterField = new Text(projectSelectorComposite,
+        SWT.BORDER | SWT.SEARCH | SWT.ICON_SEARCH | SWT.ICON_CANCEL);
+
+    projectSelector = new ProjectSelector(projectSelectorComposite);
+    projectSelector.addSelectionChangedListener(new ISelectionChangedListener() {
+      @Override
+      public void selectionChanged(SelectionChangedEvent event) {
+        if (!initializingUiValues) {
+          boolean emptySelection = projectSelector.getSelectProjectId().isEmpty();
+          boolean savedIdAvailable = projectSelector.isProjectIdAvailable(gcpProjectIdModel);
+          if (!emptySelection || savedIdAvailable) {
+            gcpProjectIdModel = projectSelector.getSelectProjectId();
+            updateLaunchConfigurationDialog();
+          }
+        }
+      }
+    });
+
+    filterField.setMessage(Messages.getString("project.filter.hint")); //$NON-NLS-1$
+    filterField.addModifyListener(new ModifyListener() {
+      @Override
+      public void modifyText(ModifyEvent event) {
+        projectSelector.setFilter(filterField.getText());
+      }
+    });
+
+    // Service key row
+    new Label(composite, SWT.LEAD).setText(Messages.getString("label.service.key")); //$NON-NLS-1$
+    serviceKeyInput = new Text(composite, SWT.BORDER);
+    serviceKeyInput.addModifyListener(new ModifyListener() {
+      @Override
+      public void modifyText(ModifyEvent event) {
+        updateLaunchConfigurationDialog();
+      }
+    });
+    Button browse = new Button(composite, SWT.NONE);
+    browse.setText(Messages.getString("button.browse")); //$NON-NLS-1$
+
+    GridDataFactory.fillDefaults().align(SWT.BEGINNING, SWT.TOP).applyTo(projectLabel);
+    GridDataFactory.fillDefaults().span(2, 1).applyTo(accountSelector);
+    GridLayoutFactory.swtDefaults().numColumns(3).generateLayout(composite);
+
+    GridDataFactory.fillDefaults().span(2, 1).applyTo(projectSelectorComposite);
+    GridDataFactory.fillDefaults().applyTo(filterField);
+    GridDataFactory.fillDefaults().grab(true, false).hint(SWT.DEFAULT, 200)
+        .applyTo(projectSelector);
+    GridLayoutFactory.fillDefaults().spacing(0, 0).generateLayout(projectSelectorComposite);
+
+    setControl(composite);
+
+    gcpIcon = SharedImages.GCP_IMAGE_DESCRIPTOR.createImage();
+    composite.addDisposeListener(new ImageDisposer(gcpIcon));
```

Comment:
You can do this by overriding `AbstractLaunchConfigurationTab.dispose()`

Reply:
Turns out, the tab lifecycle doesn't seem to match SWT lifecycle here.

Reply:
What does this mean — what is the effect? Because with this `ImageDisposer`, `gcpIcon` is still left pointing at an invalid object.

Reply:
Did some more testing. Looks like the tab's `dispose()` is matched with class instantiation, which is different that SWT's `dispose()` is matched with `createControl()`. I finally decided to make `gcpIcon` instance-scoped; `gcpIcon` is now a `final` field, created only once when the class is created.